cardano-ledger-core/src/Cardano/Ledger/BaseTypes.hs

```diff
@@ -103,54 +108,81 @@ type FixedPoint = Digits34
 fpPrecision :: FixedPoint
 fpPrecision = (10 :: FixedPoint) ^ (34 :: Integer)
 
+-- | Maximum precision possible for unit interval when backed by a 64bit
+type UnitScale = 19
+
+unitScale :: Int
+unitScale = fromInteger (natVal (Proxy :: Proxy UnitScale))
+
+newtype RawUnit
+  = RawUnit Word64
+  deriving newtype (Show, Eq, Ord, Num, Real, Enum, Integral, NFData)
+
+instance Bounded RawUnit where
+  minBound = 0
+  maxBound = RawUnit (10 ^ unitScale)
+
 -- | Type to represent a value in the unit interval [0; 1]
-newtype UnitInterval = UnsafeUnitInterval (Ratio Word64)
-  deriving (Show, Ord, Eq, Generic)
-  deriving newtype (NoThunks, NFData)
+newtype UnitInterval = UnitInterval
+  { unitDecimal :: Decimal RoundHalfEven UnitScale RawUnit
+  }
+  deriving (Generic)
+  deriving newtype (Show, Ord, Eq, Bounded, NFData)
+
+instance Default UnitInterval where
+  def = minBound
+
+instance NoThunks UnitInterval where
+  noThunks ctx = noThunks ctx . coerce @_ @Word64
+  wNoThunks ctx = wNoThunks ctx . coerce @_ @Word64
 
+-- We are using CBOR serialization through `Ratio Word64` for hisotrical reasons. It would
+-- be better to use Word64 directly: `toCBOR . coerce @_ @Word64`
```

Comment:
We should check where we're using this - if it's not being serialised in blocks, we can get away with changing the serialisation so long as the deserialiser can handle both.

Reply:
That is an interesting idea. I can investigate it further. How would we go about CDDL spec change?
Either way if it will work or not I'd suggest that we make that update in a separate PR, since I think this one is already big enough.